File: src/levels/L24-un-solo-pago/factory.rb
# Factory Method para crear un único tipo de pago: una jerarquía de creadores con un solo producto.
IVA = 0.16

class PaymentFactory
  # region: PaymentFactory#create
  def create = CardPayment.new
  # endregion
end

class CardPayment
  def charge(total) = CardTerminal.charge(total)
end

class Cashier
  def initialize(barista, factory)
    @barista = barista
    @factory = factory
  end

  # region: Cashier#charge
  def charge(order)
    @factory.create.charge(order.price * (1 + IVA))
    @barista.prepare(order)
  end
  # endregion
end

# region: wiring
cashier = Cashier.new(Barista.new(Counter.new), PaymentFactory.new)
OrderTaker.new(cashier).take(:latte, 50)
# endregion

File: src/levels/L24-un-solo-pago/ts/factory.ts
// Factory Method para crear un único tipo de pago: una jerarquía de creadores con un solo producto.
const IVA = 0.16

class PaymentFactory {
  // region: PaymentFactory#create
  create() { return new CardPayment() }
  // endregion
}

class CardPayment {
  charge(total: number) { CardTerminal.charge(total) }
}

class Cashier {
  constructor(
    private barista: Barista,
    private factory: PaymentFactory,
  ) {}

  // region: Cashier#charge
  charge(order: Order) {
    this.factory.create().charge(order.price * (1 + IVA))
    this.barista.prepare(order)
  }
  // endregion
}

// region: wiring
const cashier = new Cashier(new Barista(new Counter()), new PaymentFactory())
new OrderTaker(cashier).take('latte', 50)
// endregion
